ledger/participant-integration-api/src/main/scala/platform/store/completions/PagedCompletionsReaderWithCache.scala

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.completions
+import java.util.concurrent.atomic.AtomicReference
+
+import com.daml.ledger.ApplicationId
+import com.daml.ledger.api.v1.command_completion_service.CompletionStreamResponse
+import com.daml.ledger.participant.state.v1.Offset
+import com.daml.lf.data.Ref
+import com.daml.lf.data.Ref.Party
+import com.daml.logging.LoggingContext
+import com.daml.platform.store.dao.CommandCompletionsTable.CompletionStreamResponseWithParties
+
+import scala.collection.SortedMap
+import scala.concurrent.{ExecutionContext, Future}
+
+/** Completions reader implementation that caches maxItems of newest completions in memory.
+  * Under concurrent access, requests that miss the cache will be fetching data from database sequentially -
+  * only one request at the time can be executed on database for completions newer than last cached offset.
+  * @param completionsDao DAO object responsible for fetching completions from datastore
+  * @param maxItems maximum amount of completions stored in in-memory cache
+  * @param executionContext context in which are run async operations
+  */
+class PagedCompletionsReaderWithCache(completionsDao: CompletionsDao, maxItems: Int)(implicit
+    executionContext: ExecutionContext
+) extends PagedCompletionsReader {
+
+  protected val cacheRef: AtomicReference[RangeCache[CompletionStreamResponseWithParties]] =
+    new AtomicReference(RangeCache.empty[CompletionStreamResponseWithParties](maxItems))
+
+  private val pendingRequestRef: AtomicReference[Future[Unit]] =
+    new AtomicReference[Future[Unit]](Future.unit)
+
+  /** Returns completions stream filtered based on parameters.
+    * Newest completions read from database are added to cache
+    *
+    * @param startExclusive start offset
+    * @param endInclusive end offset
+    * @param applicationId completions application id
+    * @param parties required completions submitters
+    * @param loggingContext logging context
+    * @return
+    */
+  override def getCompletionsPage(
+      startExclusive: Offset,
+      endInclusive: Offset,
+      applicationId: ApplicationId,
+      parties: Set[Party],
+  )(implicit
+      loggingContext: LoggingContext
+  ): Future[Seq[(Offset, CompletionStreamResponse)]] = {
+    val inMemCache = cacheRef.get()
+    val requestedRange = Range(startExclusive, endInclusive)
+    val historicCompletionsFuture =
+      calculateHistoricRangeToFetch(inMemCache.range, requestedRange)
+        .map(fetchHistoric(_, applicationId, parties))
+        .getOrElse(futureEmptyList)
+
+    val futureCompletionsFuture =
+      calculateFutureRangeToFetch(inMemCache.range, requestedRange)
+        .map(fetchFuture(_, applicationId, parties))
+        .getOrElse(futureEmptyList)
+
+    val filteredCache = inMemCache
```

Comment:
Naming: this is not a `filteredCache` anymore is it? It's type is the same as the types of the `historicCompletions` and `futureCompletions` that you use below. So `cachedCompletions` would match well wouldn't it?